Closes #2, closes #4, closes #5 Added authorization manager

nikola-mladenovic opened the pull request:
@cneijenhuis @lauraluiz - We have included here all 3 issues regarding authorization handling.

cneijenhuis commented on Source/AuthManager.swift:
```diff
@@ -0,0 +1,225 @@
+    private let kAuthAccessTokenKey = "com.commercetools.authAccessTokenKey"
+
+    /// The key used for storing rrefresh token.
+    private let kAuthRrefreshTokenKey = "com.commercetools.authRrefreshTokenKey"
```
Typos: `rr`efresh / `Rr`efresh

cneijenhuis commented on Source/AuthManager.swift:
```diff
@@ -0,0 +1,225 @@
+            if let accessToken = responseDict["access_token"] as? String, expiresIn = responseDict["expires_in"] as? Double {
+                completionHandler(accessToken, nil)
+                self.accessToken = accessToken
+                // Subtracting 10 seconds from the valid period to compensate for the latency
```
We typically take a larger time off the theoretical validity period to be safe, as there are many things that can go wrong (e.g. clocks not synchronized).
I'd choose at least 5 minutes (I think in the tickets I wrote an hour, but that is maybe a bit excessive).

nikola-mladenovic replied:
Will subtracting 10 minutes, hopefully that will be sufficient :)

cneijenhuis replied:
Great, I edited the stories :+1:

cneijenhuis commented on CommercetoolsTests/AuthManagerTests.swift:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        waitForExpectationsWithTimeout(10, handler: nil)
+    }
```
A `testUserLogout()` would be great, too :+1: 

Also, can we test the error case (user doesn't exist)?

nikola-mladenovic replied:
Absolutely - since we don't perform any validation on the SDK level, I thought not to test API logic, but we could definitely add some tests for incorrect user / password and similar scenarios. Just let me know-

cneijenhuis replied:
I was thinking about testing the "error processing" branch of the SDK as well. One scenario is enough for that.

cneijenhuis commented on CommercetoolsTests/AuthManagerTests.swift:
```diff
@@ -0,0 +1,82 @@
+
+        waitForExpectationsWithTimeout(10, handler: nil)
+    }
+    
```
Is there a convenient way to test that `storeTokens()`/`loadTokens()` work?

Also, if we have that, we could maybe test the refresh flow by overwriting the validity date and the auth manager would have to use the refresh token after loading.

nikola-mladenovic replied:
The way it's currently implemented - no. But we could extract storing / loading tokens into separate object, which we could then test. I personally think we shouldn't test storing and loading objects into user defaults (or keychain, in case we migrate to that option in future), because that's part of the Foundation, not our implementation.

On the other hand, deciding to have another object handling storing / loading tokens would provide us with the ability to test refreshing, and in my opinion that would be a valid reason :)

cneijenhuis commented:
One other thing, in #4 I wrote

> The SDK should offer an API to tell whether it is currently using a password flow token or a client_credentials flow token (and be extensible for other flows, so not a Boolean :wink: ).

I think this is useful if an app needs to decide whether it renders a login/register button or other actions that are dependent on whether a user is logged in/there is an anonymous cart/it's just client_credentials.

cneijenhuis commented:
Cool, looks good so far, only small issues :wink:

cneijenhuis commented on Source/AuthManager.swift:
```diff
@@ -0,0 +1,225 @@
+    private let kAuthAccessTokenKey = "com.commercetools.authAccessTokenKey"
+
+    /// The key used for storing refresh token.
+    private let kAuthRrefreshTokenKey = "com.commercetools.authRefreshTokenKey"
```
`Rr` in `kAuthRrefreshTokenKey` :wink:

lauraluiz commented:
👍  on my side

nikola-mladenovic commented:
> I think this is useful if an app needs to decide whether it renders a login/register button or other actions that are dependent on whether a user is logged in/there is an anonymous cart/it's just client_credentials.

In order to accommodate this need, I've added public property `state` to the auth manager, which actually represents the current state, and in case we have more in the future, it can be easily added to the `AuthState` enum.
Also, I've updated current tests to cover state checking as well.

cneijenhuis commented on Source/AuthManager.swift:
```diff
@@ -10,11 +10,27 @@ import Alamofire
+        case Anonymous
+        case Authenticated
+        case Undefined
+    }
```
I discussed a bit with @stmeissner. I think `Anonymous` is ok now, but once we have the (refresh) token for the anonymous checkout, it becomes less clear. I propose the following (feedback welcome):

```swift
/**
    Enum used to specify current token state.

    - CustomerToken: Auth manager is handling tokens for a logged in customer.
    - AnonymousToken: Auth manager is handling tokens for an anonymous customer.
    - PlainToken: Auth manager is handling a token without an associated customer.
    - NoToken: Auth manager does not have a token (i.e. because Commercetools configuration is not valid)
*/

public enum TokenState: Int {
    case CustomerToken
    case AnonymousToken
    case PlainToken
    case NoToken
}
```

`CustomerToken` = `Authenticated`
`PlainToken` = `Anonymous`
`NoToken` = `Undefined`
(Note that `AnonymousToken` is not used yet).

What do you think?

nikola-mladenovic replied:
Please do correct me if I'm wrong, but my impression is that once we have anonymous checkout and anonymous refresh token in place, we won't have a `PlainToken` use case?

I do agree with the feedback, just let me know if you think it makes sense to remove `PlainToken` case? If we do that, for now our `AnonymousToken` description will change once we have anonymous customer support on the API.

On the other hand, if you have something in mind, which would justify having both `PlainToken` and `AnonymousToken` let's go with both?

cneijenhuis replied:
Yes, because the anonymous checkout should be optional. Some of our clients want to require accounts and do not allow anonymous checkout (especially in the B2B space).

The `AnonymousToken` will get a new OAuth scope to allow for that.

nikola-mladenovic replied:
Ah, now I do understand. Thanks for clarifying :wink:

cneijenhuis commented:
As always, naming is the hardest part :laughing:

nikola-mladenovic commented:
Added token store, keeping tokens in keychain, updated state enums, added test cases for logout and refresh token.

cneijenhuis commented:
Awesome, LGTM :+1:
